Sources/SourceKitLSP/Rename.swift

```diff
@@ -284,78 +291,438 @@ private extension RenameLocation.Usage {
+      isObjectiveCSelector: self.isObjectiveCSelector
+    )
+  }
+
+  /// Get the name of the symbol as it should be used to perform a rename in clang files.
+  ///
+  /// For example, this performs Swift function to Objective-C selector translation or strips parentheses for Swift
+  /// functions imported into C.
+  func clangName() async throws -> String {
+    switch definitionLanguageService {
+    case let swiftLanguageService as SwiftLanguageServer:
+      if swiftToClangTranslationTask == nil {
+        swiftToClangTranslationTask = Task {
+          try await swiftLanguageService.translateSwiftNameToClang(
+            at: definitionPosition,
+            in: definitionDocumentUri,
+            name: CompoundDeclName(definitionName)
+          )
+        }
+      }
+      return try await swiftToClangTranslationTask!.value
+    case is ClangLanguageServerShim:
+      return definitionName
+    default:
+      throw ResponseError.unknown("Cannot rename symbol because it is defined in an unknown language")
+    }
+  }
+
+  /// Get the name of the symbol as it should be used to perform a rename in Swift files.
+  ///
+  /// This runs the clang importer to translates Objective-C selectors to Swift functions.
+  ///
+  /// - Note: The symbol translation from clang to Swift requires a Swift location at which the clang symbol is
+  ///   (probably so that sourcekitd can access the necessary AST context and get the clang importer to translate the
+  ///   name). But the output does not actually matter on the specific position and URI. We can thus cache the result.
+  ///   Ideally, sourcekitd wouldn't need a position and URI to translate a clang name to Swift.
+  func swiftName(
+    at position: Position,
+    in snapshot: DocumentSnapshot,
+    languageService: SwiftLanguageServer
+  ) async throws -> String {
+    switch definitionLanguageService {
+    case is SwiftLanguageServer:
+      return definitionName
+    case is ClangLanguageServerShim:
+      if clangToSwiftTranslationTask == nil {
+        clangToSwiftTranslationTask = Task {
+          try await languageService.translateClangNameToSwift(
+            at: position,
+            in: snapshot,
+            isObjectiveCSelector: isObjectiveCSelector,
+            name: definitionName
+          )
+        }
+      }
+      return try await clangToSwiftTranslationTask!.value
+    default:
+      throw ResponseError.unknown("Cannot rename symbol because it is defined in an unknown language")
+    }
+  }
+}
+
 // MARK: - SourceKitServer
 
 extension SourceKitServer {
+  private func getTranslatableName(
+    forUsr usr: String,
+    workspace: Workspace,
+    index: IndexStoreDB
+  ) async throws -> (name: TranslatableName, language: Language)? {
+    guard let definitionSymbol = index.occurrences(ofUSR: usr, roles: [.definition]).only else {
+      logger.error("Multiple or no definitions for \(usr) found")
+      return nil
+    }
```

Comment:
Multiple reasons:
- Yes, language is one of them
- The entire design currently centers around `ToolchainLanguageService.rename` returning a list of local declarations + the symbol’s USR of the symbol but the definition request would need a location of a reference. Sure we could pass it down but it seems cleaner to just stay in the USR-based realm here.
- It’s less work to just do an index lookup than a definition request (which would need to go through sourcekitd/clangd) and since we need the index for cross-file rename anyway, we wouldn’t benefit from the fallbacks that the definition request can provide.

Reply:
Fair enough, I suppose I was thinking that the definition request may be less out of date but... that's not true at all in the current design (they are both as up to date as each other). And once we have background indexing, we would have to make sure the index is up to date anyway 👍